appc/appc.go

```diff
@@ -0,0 +1,328 @@
+	dnsFailures    expvar.Int
+	tcpConns       expvar.Int
+	sniConns       expvar.Int
+	unhandledConns expvar.Int
+}
+
+// InstallMetrics publishes metrics about the operation of the app connector.
+// This method must only be called once.
+func (s *Server) InstallMetrics() {
+	stats := new(metrics.Set)
+
+	stats.Set("tls_sessions", &s.m.sniConns)
+	clientmetric.NewCounterFunc("sniproxy_tls_sessions", s.m.sniConns.Value)
+
+	stats.Set("tcp_sessions", &s.m.tcpConns)
+	clientmetric.NewCounterFunc("sniproxy_tcp_sessions", s.m.tcpConns.Value)
+
+	stats.Set("dns_responses", &s.m.dnsResponses)
+	clientmetric.NewCounterFunc("sniproxy_dns_responses", s.m.dnsResponses.Value)
+
+	stats.Set("dns_failed", &s.m.dnsFailures)
+	clientmetric.NewCounterFunc("sniproxy_dns_failed", s.m.dnsFailures.Value)
+
+	expvar.Publish("sniproxy", stats)
+}
+
+// Configure applies the provided configuration to the app connector.
+func (s *Server) Configure(cfg *appctype.AppConnectorConfig) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.connectors = makeConnectorsFromConfig(cfg)
+}
+
+// HandleTCPFlow implements tsnet.FallbackTCPHandler.
+func (s *Server) HandleTCPFlow(src, dst netip.AddrPort) (handler func(net.Conn), intercept bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, c := range s.connectors {
+		if handler, intercept := c.handleTCPFlow(src, dst, &s.m); intercept {
+			return handler, intercept
+		}
+	}
+
+	return nil, false
+}
+
+// HandleDNS handles a DNS request to the app connector.
+func (s *Server) HandleDNS(c nettype.ConnPacketConn) {
+	defer c.Close()
+	c.SetReadDeadline(time.Now().Add(5 * time.Second))
+	buf := make([]byte, 1500)
+	n, err := c.Read(buf)
+	if err != nil {
+		log.Printf("HandleDNS: read failed: %v\n ", err)
+		s.m.dnsFailures.Add(1)
+		return
+	}
+
+	addrPortStr := c.LocalAddr().String()
+	host, _, err := net.SplitHostPort(addrPortStr)
+	if err != nil {
+		log.Printf("HandleDNS: bogus addrPort %q", addrPortStr)
+		s.m.dnsFailures.Add(1)
+		return
+	}
+	localAddr, err := netip.ParseAddr(host)
+	if err != nil {
+		log.Printf("HandleDNS: bogus local address %q", host)
+		s.m.dnsFailures.Add(1)
+		return
+	}
+
+	var msg dnsmessage.Message
+	err = msg.Unpack(buf[:n])
+	if err != nil {
+		log.Printf("HandleDNS: dnsmessage unpack failed: %v\n ", err)
+		s.m.dnsFailures.Add(1)
+		return
+	}
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for _, connector := range s.connectors {
+		resp, err := connector.handleDNS(&msg, localAddr)
+		if err != nil {
+			log.Printf("HandleDNS: connector handling failed: %v\n", err)
+			s.m.dnsFailures.Add(1)
+			return
+		}
+		if len(resp) > 0 {
+			// This connector handled the DNS request
+			_, err = c.Write(resp)
+			if err != nil {
+				log.Printf("HandleDNS: write failed: %v\n", err)
+				s.m.dnsFailures.Add(1)
+				return
+			}
+
+			s.m.dnsResponses.Add(1)
+			return
+		}
+	}
+}
+
+// connector describes a logical collection of
+// services which need to be proxied.
+type connector struct {
+	Handlers map[target]handler
```

Comment:
does handlers need to be exported?

Reply:
It just makes tests a lot easier. We aren't accidentally exposing something we shouldnt, because the underlying types arent exported.